src/balancer/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+#![warn(clippy::clone_on_ref_ptr)]
+#![warn(clippy::no_effect)]
+#![warn(clippy::unnecessary_unwrap)]
+#![warn(clippy::dbg_macro)]
+#![warn(clippy::todo)]
+#![warn(clippy::wildcard_dependencies)]
+#![warn(clippy::zero_prefixed_literal)]
+#![warn(clippy::borrowed_box)]
+#![warn(clippy::deref_addrof)]
+#![warn(clippy::double_must_use)]
+#![warn(clippy::double_parens)]
+#![warn(clippy::extra_unused_lifetimes)]
+#![warn(clippy::needless_borrow)]
+#![warn(clippy::needless_question_mark)]
+#![warn(clippy::needless_return)]
+#![warn(clippy::redundant_pattern)]
+#![warn(clippy::redundant_slicing)]
+#![warn(clippy::redundant_static_lifetimes)]
+#![warn(clippy::single_component_path_imports)]
+#![warn(clippy::unnecessary_cast)]
+#![warn(clippy::useless_asref)]
+#![warn(clippy::useless_conversion)]
+#![warn(clippy::builtin_type_shadow)]
+#![warn(clippy::duplicate_underscore_argument)]
+#![warn(clippy::double_neg)]
+#![warn(clippy::unnecessary_mut_passed)]
+#![warn(clippy::wildcard_in_or_patterns)]
+#![warn(clippy::crosspointer_transmute)]
+#![warn(clippy::excessive_precision)]
+#![warn(clippy::overflow_check_conditional)]
+#![warn(clippy::as_conversions)]
+#![warn(clippy::match_overlapping_arm)]
+#![warn(clippy::zero_divided_by_zero)]
+#![warn(clippy::must_use_unit)]
+#![warn(clippy::suspicious_assignment_formatting)]
+#![warn(clippy::suspicious_else_formatting)]
+#![warn(clippy::suspicious_unary_op_formatting)]
+#![warn(clippy::mut_mutex_lock)]
+#![warn(clippy::print_literal)]
+#![warn(clippy::same_item_push)]
+#![warn(clippy::useless_format)]
+#![warn(clippy::write_literal)]
+#![warn(clippy::redundant_closure)]
+#![warn(clippy::redundant_closure_call)]
+#![warn(clippy::unnecessary_lazy_evaluations)]
+#![warn(clippy::partialeq_ne_impl)]
+#![warn(clippy::redundant_field_names)]
+#![warn(clippy::transmutes_expressible_as_ptr_casts)]
+#![warn(clippy::unused_async)]
+#![warn(clippy::disallowed_methods)]
+#![warn(clippy::disallowed_macros)]
+#![warn(clippy::disallowed_types)]
+#![warn(clippy::from_over_into)]
+// END LINT CONFIG
+
+mod codec;
+
+use std::collections::BTreeMap;
+use std::net::SocketAddr;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::time::Instant;
+
+use bytes::BytesMut;
+use mz_build_info::{build_info, BuildInfo};
+use mz_frontegg_auth::Authentication as FronteggAuthentication;
+use mz_ore::metrics::{ComputedGauge, MetricsRegistry};
+use mz_ore::netio::AsyncReady;
+use mz_ore::server::{listen, TlsCertConfig, TlsConfig, TlsMode};
+use mz_ore::{metric, task};
+use mz_pgwire_common::{
+    decode_startup, Conn, ErrorResponse, FrontendMessage, FrontendStartupMessage,
+    ACCEPT_SSL_ENCRYPTION, REJECT_ENCRYPTION, VERSION_3,
+};
+use openssl::ssl::Ssl;
+use semver::Version;
+use tokio::io::{self, AsyncRead, AsyncWrite, AsyncWriteExt};
+use tokio::net::TcpStream;
+use tokio_openssl::SslStream;
+use tokio_postgres::error::SqlState;
+use tracing::{error, warn};
+
+use crate::codec::{BackendMessage, FramedConn};
+
```

Comment:
Can you add a module comment here explaining at a high level what a "balancer" is? Some questions that would be great to answer:

1. Do we run multiple balancers in parallel, are they stateless?
2. Are they load balancers, reverse proxies, both, neither?
3. Do we run a balancer per-customer, or a few for all of our customers?
4. Roughly where does a balancer fall in the stack of adapter and clusterd?

Reply:
Done. I was not able to figure out how to get this comment to show up in the generated docs, but I wrote some words!